### PR TITLE
Validate NodeConfig

### PR DESCRIPTION
**Pull request by adambabik**
Related to issue #149.

### Changes
* exposed `ValidateNodeConfig()`,
* extended `NodeConfig` interface with `Validate()` method,
* added `validate` tag to some `NodeConfig` fields.

### Validated fields
* `NodeConfig.NetworkID` - required, must be in a set of valid network ids,
* `NodeConfig.DataDir` - required,
* `NodeConfig.NodeKeyFile` - required (**disabled** as many tests failed so I am not sure it's actually required field...),
* `NodeConfig.Name` - if value present, then does not contain `/`.

### Error formatting

Errors, as a string, will have a format:
```
Key: 'NodeConfig.DataDir' Error:Field validation for 'DataDir' failed on the 'required' tag
Key: 'NodeConfig.NodeKeyFile' Error:Field validation for 'NodeKeyFile' failed on the 'required' tag
```
i.e., multiple errors can be returned, each one in a new line.

Todo:
* [ ] better error formatting (some help from ClojureTeam is required),
* [x] add validation for the rest of the fields.

### CODE REVIEW

**Comment by tiabc** [comment]
Thanks, do you have access to create branches? Could you move your pull request from a forked repository to a separate branch in `status-go`?

**Comment by adambabik** [comment]
@tiabc yes I have and this PR comes from a branch in `status-go`. The branch name is `feat/validate-json-config-#149`.

**Comment by tiabc** [comment]
@adambabik, ah, sorry, I've mislooked it half-awake.

File: geth/params/config_test.go [hunk 1]
```diff
@@ -8,6 +8,8 @@ import (
 	"strings"
 	"testing"
 
+	validator "gopkg.in/go-playground/validator.v9"
```
Comment (tiabc) [inline geth/params/config_test.go]: Is there a reason to add an alias?

Comment (adambabik) [inline geth/params/config_test.go]: Not really. It must have been my editor. I will remove it.

**Comment by adambabik** [comment]
@rasom please check "Error formatting" section in the description. This is a little bit different from the previous output. I am not sure how the errors will be presented in the app. Let me know if error formatting could be better.

Another thing is "Validated fields". I listed all fields that are validated now. Do you have any suggestions which other fields should be validated? cc @tiabc

**Comment by tiabc** [comment]
You seem to assume that only fields with values coming from `params.NewNodeConfig` need validating. This is not so due to: https://github.com/status-im/status-go/blob/95c0f4ac05ae61f1d5e105afb1058a642f0d0503/cmd/statusd/library.go#L29-L29

The react application seems to use the `StartNode` signal to start our application and it passes JSON which also needs validating because it's passed from an untrusted source. So all the other fields in our configuration need some sanity checks. @rasom, please, correct me if I'm wrong.

**Comment by adambabik** [comment]
@tiabc the case is that most of these fields have default values or accept empty values. That's why I asked about a list of fields that need special validation. 

Moreover, if some fields need to be validate just because status-go is used within the React app, then the React app itself should do that validation. What I mean here is that e.g. `NodeKeyFile` can be empty and status-go will be fine with it, but the React app won't, then the React app should do the validation.

**Comment by tiabc** [comment]
Yes, I understand.

As I know, status-react calls `GenerateConfig`, stores it somewhere and then just feeds it to `status-go` upon start (again, I may be wrong here, @rasom, please, correct me if so). Meanwhile, a user can edit the configuration and break something, even if there're some defaults. From my point of view, `status-react` has nothing to do with the config, it just needs to ensure this config is valid and the only component which can ensure it is `status-go` so it's our responsibility to tell bad configuration from the valid.

This task is indeed crosses the borders of our applications in a sense that you need to determine which parameters are essential for status-react operation yourself. They may not even know which.

**Comment by adambabik** [comment]
Let's wait for the Clojure Team to comment on this. 

I checked `status-react` source code and it's exactly as you described, i.e. `GenerateConfig()` is called and then the returned config is used to call `StartNode()`. There is no way to update config after that. So, `status-react` would need to first stop the node and start it again with a different config. Also, I can't find how to change the config in the UI (using `develop` branch).

**Comment by tiabc** [comment]
@rasom:
> so in overall, what we need from status-go side: we want to be sure that when user passes some configuration (arbitrary, imagine that user will have control over all parameters) that configuration will be validated before applying and the next problems should be detected:
> 1) absence of required parameters
> 2) wrong types (bool instead of string, etc)
> 3) wrong parameters (keys)
> 4) wrong parameters values (like we have some enum for NetworkId, and user passes wrong value or so)
> 5) whatever else that makes sense, like detecting of mutually exclusive parameters (if such exist)

**Comment by rasom** [comment]
Regarding formatting:
the best choice for us will be some JSON like this (or maybe with a bit different structure if you want, but similar) 
```json
{
    "errors":
        [
            {
                "parameter": "some_parameter",
                "errors": [
                    {
                        "code": 0,
                        "message": "some description"
                    },
                    {
                        "code": 1,
                        "message": "another description"
                    }]
            },
            {
                "parameter": "some.nested_key",
                "errors": [
                    {
                        "code": 0,
                        "message": "some description"
                    }
                ]
                ...
            }],
    ...
}
```

**Comment by adambabik** [comment]
Work will be continued in branch `feature/validate-json-config-#149` as a base branch should be `develop` now.

**Pull request closed without merging**